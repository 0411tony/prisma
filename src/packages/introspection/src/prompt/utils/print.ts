import chalk from 'chalk'
import { ConnectorType } from '@prisma/generator-helper'

export function printError(text) {
  return chalk.bold.bgRed(' ERROR ') + ' ' + chalk.red(text)
}

export function printFix(text) {
  return chalk.bgWhite.black.bold(' FIX ') + ' ' + text
}

const beautifyMap = {
  javascript: 'JavaScript',
  typescript: 'TypeScript',
}

export function beautifyLanguage(language: string): string {
  return beautifyMap[language] || language
}

export function prettyDb(type: ConnectorType) {
  switch (type) {
    case 'mysql': {
      return 'MySQL'
    }
    case 'postgresql': {
      return 'PostgreSQL'
    }
  }
  return type
}
