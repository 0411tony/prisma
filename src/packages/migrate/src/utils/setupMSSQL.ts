import fs from 'fs'
import path from 'path'
import mssql from 'mssql'

export type SetupParams = {
  connectionString: string
  dirname: string
}

export async function setupMSSQL(options: SetupParams): Promise<void> {
  const { connectionString } = options
  const { dirname } = options

  let schema = `
    IF NOT EXISTS(SELECT * FROM sys.databases WHERE name = 'tests-migrate')
    BEGIN
      CREATE DATABASE [tests-migrate]
    END
    IF NOT EXISTS(SELECT * FROM sys.databases WHERE name = 'tests-migrate-shadowdb')
    BEGIN
      CREATE DATABASE [tests-migrate-shadowdb]
    END

    USE [tests-migrate]
    `
  if (dirname !== '') {
    schema += fs.readFileSync(path.join(dirname, 'setup.sql'), 'utf-8')
  }

  const connectionPool = new mssql.ConnectionPool(connectionString)
  const connection = await connectionPool.connect()

  await connection.query(schema)
  void connection.close()
}

export async function tearDownMSSQL(options: SetupParams) {
  const { connectionString } = options
  const connectionPool = new mssql.ConnectionPool(connectionString)
  const connection = await connectionPool.connect()

  await connection.query(`
  USE "tests-migrate"

  -- Thanks to https://blog.ramondeklein.nl/2016/08/10/clean-up-sql-servers-master-database/
  DECLARE @statement nvarchar(250)

  -- Kill all foreign keys
  DECLARE ForeignKeyCursor CURSOR LOCAL FORWARD_ONLY
  FOR SELECT fk.name, tab.name
      FROM sys.objects fk
      INNER JOIN sys.objects tab ON fk.parent_object_id = tab.object_id
      WHERE fk.TYPE = 'F'
  
  DECLARE @keyName sysname
  DECLARE @tableName sysname
  
  OPEN ForeignKeyCursor
  FETCH NEXT FROM ForeignKeyCursor INTO @keyName, @tableName
  WHILE @@FETCH_STATUS = 0
  BEGIN
      SET @statement = 'ALTER TABLE [' + @tableName + '] DROP CONSTRAINT [' + @keyName + ']'
      EXEC sp_executeSql @statement
  
      FETCH NEXT FROM ForeignKeyCursor INTO @keyName, @tableName
  END
  CLOSE ForeignKeyCursor
  DEALLOCATE ForeignKeyCursor
  
  -- Remove all objects in this database
  DECLARE ObjectCursor CURSOR LOCAL FORWARD_ONLY
  FOR SELECT name, type FROM sys.objects
  DECLARE @name sysname
  DECLARE @type nvarchar(10)
  
  OPEN ObjectCursor
  FETCH NEXT FROM ObjectCursor INTO @name, @type
  WHILE @@FETCH_STATUS = 0
  BEGIN
      SET @statement =
          CASE @type
          WHEN 'FN' THEN 'DROP FUNCTION [' + @name + ']'
          WHEN 'IF' THEN 'DROP FUNCTION [' + @name + ']'
          WHEN 'TF' THEN 'DROP FUNCTION [' + @name + ']'
          WHEN 'P' THEN 'DROP PROCEDURE [' + @name + ']'
          WHEN 'U' THEN 'DROP TABLE [' + @name + ']'
          WHEN 'V' THEN 'DROP VIEW [' + @name + ']'
          ELSE null
      END
  
      IF @statement IS NOT NULL
          EXEC sp_executeSql @statement
  
      FETCH NEXT FROM ObjectCursor INTO @name, @type
  END
  CLOSE ObjectCursor
  DEALLOCATE ObjectCursor
  
  -- Remove all user defined types in this database
  DECLARE UdtCursor CURSOR LOCAL FORWARD_ONLY
  FOR SELECT name FROM sys.Types WHERE is_user_defined = 1
  DECLARE @udtName sysname
  
  OPEN UdtCursor
  FETCH NEXT FROM UdtCursor INTO @udtName
  WHILE @@FETCH_STATUS = 0
  BEGIN
      SET @statement = 'DROP TYPE [' + @udtName + ']'
      EXEC sp_executeSql @statement
  
      FETCH NEXT FROM UdtCursor INTO @udtName
  END
  CLOSE UdtCursor
  DEALLOCATE UdtCursor
`)
  void connection.close()
}
