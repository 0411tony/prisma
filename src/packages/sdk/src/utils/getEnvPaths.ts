import debugLib from 'debug'
import findUp from 'find-up'
import path from 'path'
import { getSchemaPathFromPackageJsonSync } from '../cli/getSchema'
import { exists } from './tryLoadEnvs'

const debug = debugLib('loadEnv')
/**
 *  1. Search in project root
 *  1. Schema
 *    1. Search location from schema arg --schema
 *    1. Search location from pkgJSON `"prisma": {"schema": "/path/to/schema.prisma"}`
 *    1. Search default location `./prisma/.env`
 *    1. Search cwd `./.env`
 *
 * @returns `{ rootEnvPath, schemaEnvPath }`
 */
export function getEnvPaths(
  schemaPath?: string | null,
  opts: { cwd: string } = { cwd: process.cwd() },
) {
  const rootEnvPath = getProjectRootEnvPath({ cwd: opts.cwd }) ?? null
  const schemaEnvPathFromArgs = schemaPathToEnvPath(schemaPath)
  const schemaEnvPathFromPkgJson = schemaPathToEnvPath(
    readSchemaPathFromPkgJson(),
  )
  const schemaEnvPaths = [
    schemaEnvPathFromArgs, // 1 - Check --schema directory for .env
    schemaEnvPathFromPkgJson, // 2 - Check package.json schema directory for .env
    './prisma/.env', // 3 - Check ./prisma directory for .env
    './.env', // 4 - Check cwd for .env
  ]
  const schemaEnvPath = schemaEnvPaths.find(exists)
  return { rootEnvPath, schemaEnvPath }
}


function readSchemaPathFromPkgJson(): string | null {
  try {
    return getSchemaPathFromPackageJsonSync(process.cwd())
  } catch {
    return null
  }
}

function getProjectRootEnvPath(opts: findUp.Options | undefined) {
  const pkgJsonPath = findUp.sync((dir) => {
    const pkgPath = path.join(dir, 'package.json')
    if (findUp.exists(pkgPath)) {
      try {
        let pkg = require(pkgPath)
        if (pkg['name'] !== '.prisma/client') {
          return pkgPath
        }
      } catch (e) {
        debug(e)
      }
    }
  }, opts)

  const projectRootDir = pkgJsonPath && path.dirname(pkgJsonPath)
  return projectRootDir && path.join(projectRootDir, '.env')
}
function schemaPathToEnvPath(schemaPath: string | null | undefined) {
  if (!schemaPath) return null
  return path.join(path.dirname(schemaPath), '.env')
}
