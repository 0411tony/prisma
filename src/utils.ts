import leven from 'js-levenshtein'

export type Dictionary<T> = { [key: string]: T }

export const keyBy: <T>(collection: Array<T>, iteratee: (value: T) => string) => Dictionary<T> = (
  collection,
  iteratee,
) => {
  return collection.reduce((acc, curr) => {
    acc[iteratee(curr)] = curr
    return acc
  }, {})
}

export const ScalarTypeTable = {
  String: true,
  Int: true,
  Float: true,
  Boolean: true,
  Long: true,
  DateTime: true,
  ID: true,
  UUID: true,
  Json: true,
}

export function isScalar(str: string): boolean {
  if (typeof str !== 'string') {
    return false
  }
  return ScalarTypeTable[str] || false
}

export function getSuggestion(str: string, possibilities: string[]): string | null {
  const bestMatch = possibilities.reduce(
    (acc, curr) => {
      const distance = leven(str, curr)
      if (distance < acc.distance) {
        return {
          distance,
          str: curr,
        }
      }

      return acc
    },
    {
      distance: str.length, // if the whole string would need to be replaced, it doesn't make sense
      str: null,
    },
  )

  return bestMatch.str
}
