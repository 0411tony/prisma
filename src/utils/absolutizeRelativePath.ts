import { URL } from 'url'
import path from 'path'

export function absolutizeRelativePath(filePath: string, cwd: string): string {
  const url = new URL(filePath)
  if (url.protocol !== 'file:') {
    return filePath
  }

  const joined = path.join(cwd, url.hostname)
  return `file:/${joined}`
}
