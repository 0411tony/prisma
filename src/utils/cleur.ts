'use strict'

const { FORCE_COLOR, NODE_DISABLE_COLORS, TERM } = process.env

const cleur = {
  enabled: !NODE_DISABLE_COLORS && TERM !== 'dumb' && FORCE_COLOR !== '0',

  // modifiers
  reset: init(0, 0),
  bold: init(1, 22),
  dim: init(2, 22),
  italic: init(3, 23),
  underline: init(4, 24),
  inverse: init(7, 27),
  hidden: init(8, 28),
  strikethrough: init(9, 29),

  // colors
  black: init(30, 39),
  red: init(31, 39),
  green: init(32, 39),
  yellow: init(33, 39),
  blue: init(34, 39),
  magenta: init(35, 39),
  cyan: init(36, 39),
  white: init(37, 39),
  gray: init(90, 39),
  grey: init(90, 39),

  // Bright color
  blackBright: init(90, 39),
  redBright: init(91, 39),
  greenBright: init(92, 39),
  yellowBright: init(93, 39),
  blueBright: init(94, 39),
  magentaBright: init(95, 39),
  cyanBright: init(96, 39),
  whiteBright: init(97, 3),

  // background colors
  bgBlack: init(40, 49),
  bgRed: init(41, 49),
  bgGreen: init(42, 49),
  bgYellow: init(43, 49),
  bgBlue: init(44, 49),
  bgMagenta: init(45, 49),
  bgCyan: init(46, 49),
  bgWhite: init(47, 49),
}

function run(arr, str) {
  let i = 0,
    tmp,
    beg = '',
    end = ''
  for (; i < arr.length; i++) {
    tmp = arr[i]
    beg += tmp.open
    end += tmp.close
    if (str.includes(tmp.close)) {
      str = str.replace(tmp.rgx, tmp.close + tmp.open)
    }
  }
  return beg + str + end
}

function chain(has, keys) {
  let ctx: any = { has, keys }

  ctx.reset = cleur.reset.bind(ctx)
  ctx.bold = cleur.bold.bind(ctx)
  ctx.dim = cleur.dim.bind(ctx)
  ctx.italic = cleur.italic.bind(ctx)
  ctx.underline = cleur.underline.bind(ctx)
  ctx.inverse = cleur.inverse.bind(ctx)
  ctx.hidden = cleur.hidden.bind(ctx)
  ctx.strikethrough = cleur.strikethrough.bind(ctx)

  ctx.black = cleur.black.bind(ctx)
  ctx.red = cleur.red.bind(ctx)
  ctx.green = cleur.green.bind(ctx)
  ctx.yellow = cleur.yellow.bind(ctx)
  ctx.blue = cleur.blue.bind(ctx)
  ctx.magenta = cleur.magenta.bind(ctx)
  ctx.cyan = cleur.cyan.bind(ctx)
  ctx.white = cleur.white.bind(ctx)
  ctx.gray = cleur.gray.bind(ctx)
  ctx.grey = cleur.grey.bind(ctx)

  // Bright color
  ctx.blackBright = cleur.blackBright.bind(ctx)
  ctx.redBright = cleur.redBright.bind(ctx)
  ctx.greenBright = cleur.greenBright.bind(ctx)
  ctx.yellowBright = cleur.yellowBright.bind(ctx)
  ctx.blueBright = cleur.blueBright.bind(ctx)
  ctx.magentaBright = cleur.magentaBright.bind(ctx)
  ctx.cyanBright = cleur.cyanBright.bind(ctx)
  ctx.whiteBright = cleur.whiteBright.bind(ctx)

  ctx.bgBlack = cleur.bgBlack.bind(ctx)
  ctx.bgRed = cleur.bgRed.bind(ctx)
  ctx.bgGreen = cleur.bgGreen.bind(ctx)
  ctx.bgYellow = cleur.bgYellow.bind(ctx)
  ctx.bgBlue = cleur.bgBlue.bind(ctx)
  ctx.bgMagenta = cleur.bgMagenta.bind(ctx)
  ctx.bgCyan = cleur.bgCyan.bind(ctx)
  ctx.bgWhite = cleur.bgWhite.bind(ctx)

  return ctx
}

function init(open, close) {
  let blk = {
    open: `\x1b[${open}m`,
    close: `\x1b[${close}m`,
    rgx: new RegExp(`\\x1b\\[${close}m`, 'g'),
  }
  return function(this: any, txt?: any) {
    if (this !== void 0 && this.has !== void 0) {
      this.has.includes(open) || (this.has.push(open), this.keys.push(blk))
      return txt === void 0
        ? this
        : cleur.enabled
        ? run(this.keys, txt + '')
        : txt + ''
    }
    return txt === void 0
      ? chain([open], [blk])
      : cleur.enabled
      ? run([blk], txt + '')
      : txt + ''
  }
}

export default cleur
