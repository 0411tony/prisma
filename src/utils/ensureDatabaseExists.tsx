import { getSchema } from '@prisma/cli'
import { BorderBox, DummySelectable, TabIndexProvider } from '@prisma/ink-components'
import { getConfig } from '@prisma/photon'
import ansiEscapes from 'ansi-escapes'
import chalk from 'chalk'
import { Box, Color, Instance, render } from 'ink'
import Spinner from 'ink-spinner'
import React, { useState } from 'react'
import { createDatabase } from '..'
import { canConnectToDatabase } from '../liftEngineCommands'
import { Link } from './Link'
import { DatabaseCredentials, uriToCredentials } from './uriToCredentials'
const AnySpinner: any = Spinner

export type LiftAction = 'create' | 'apply' | 'unapply' | 'dev'

export async function ensureDatabaseExists(action: LiftAction, forceCreate: boolean = false) {
  const datamodel = await getSchema()
  const config = await getConfig(datamodel)
  const activeDatasource =
    config.datasources.length === 1
      ? config.datasources[0]
      : config.datasources.find(d => d.config.enabled === 'true') || config.datasources[0]

  if (!activeDatasource) {
    throw new Error(`Couldn't find a datasource in the schema.prisma file`)
  }

  const { status } = await canConnectToDatabase(activeDatasource.url.value)
  if (status === 'Ok') {
    return
  }

  if (forceCreate) {
    await createDatabase(activeDatasource.url.value)
  } else {
    await interactivelyCreateDatabase(activeDatasource.url.value, action)
  }
}

export async function interactivelyCreateDatabase(connectionString: string, action: LiftAction): Promise<void> {
  await askToCreateDb(connectionString, action)
}

export async function askToCreateDb(connectionString: string, action: LiftAction): Promise<void> {
  return new Promise(resolve => {
    let app: Instance | undefined

    const onDone = () => {
      if (app) {
        app.unmount()
        app.waitUntilExit()
      }
      // .write as console.log introduces an unwanted linebreak here
      process.stdout.write(ansiEscapes.eraseLines(11)) // height of the dialog
      resolve()
    }

    app = render(
      <TabIndexProvider>
        <CreateDatabaseDialog connectionString={connectionString} action={action} onDone={onDone} />
      </TabIndexProvider>,
    )
  })
}

interface DialogProps {
  connectionString: string
  action: LiftAction
  onDone: () => void
}

const CreateDatabaseDialog: React.FC<DialogProps> = ({ connectionString, action, onDone }) => {
  const [creating, setCreating] = useState(false)
  async function onSelect(shouldCreate: boolean) {
    if (shouldCreate) {
      setCreating(true)
      await createDatabase(connectionString)
      setCreating(false)
      onDone()
    } else {
      process.exit(0)
    }
  }
  const credentials = uriToCredentials(connectionString)
  const dbName = credentials.database
  const dbType =
    credentials.type === 'mysql'
      ? 'MySQL'
      : credentials.type === 'postgres'
      ? 'PostgreSQL'
      : credentials.type === 'sqlite'
      ? 'Sqlite'
      : credentials.type

  const schemaWord = 'database'

  return (
    <Box flexDirection="column">
      <Box flexDirection="column">
        {action === 'dev' ? (
          <Color>
            You are trying to run <Color bold>prisma2 dev</Color> for {dbType} {schemaWord} <Color bold>{dbName}</Color>
            .
          </Color>
        ) : (
          <Color>
            You are trying to {action} a migration for {dbType} {schemaWord} <Color bold>{dbName}</Color>.
          </Color>
        )}
        <Color>
          A {schemaWord} with that name doesn't exist at <Color bold>{getDbLocation(credentials)}</Color>.
        </Color>
        <Color>Do you want to create the {schemaWord}?</Color>
      </Box>
      <BorderBox flexDirection="column" title={chalk.bold('Database options')} marginTop={1}>
        {creating ? (
          <DummySelectable tabIndex={0}>
            <Color cyan>
              <AnySpinner /> Creating {schemaWord} <Color bold>{dbName}</Color>
            </Color>
          </DummySelectable>
        ) : (
          <Link
            label="Yes"
            description={`Create new ${dbType} ${schemaWord} ${chalk.bold(dbName!)}`}
            tabIndex={0}
            onSelect={() => onSelect(true)}
          />
        )}
        <Link label="No" description={`Don't create the ${schemaWord}`} tabIndex={1} onSelect={() => onSelect(false)} />
      </BorderBox>
    </Box>
  )
}

function getDbLocation(credentials: DatabaseCredentials): string {
  if (credentials.type === 'sqlite') {
    return credentials.uri!
  }

  return `${credentials.host}:${credentials.port}`
}
