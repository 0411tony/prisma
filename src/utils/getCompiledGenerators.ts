import { CompiledGeneratorDefinition, Dictionary } from '@prisma/cli'
import path from 'path'
import { GeneratorDefinitionWithPackage } from '../types'
import { LiftEngine } from '../LiftEngine'
import { generateInThread } from '../generateInThread'

export async function getCompiledGenerators(
  cwd: string,
  datamodel: string,
  definitions: Dictionary<GeneratorDefinitionWithPackage>,
): Promise<CompiledGeneratorDefinition[]> {
  const engine = new LiftEngine({ projectDir: cwd })
  const config = await engine.getConfig({ datamodel })

  const generators = config.generators.map(g => ({
    ...g,
    output: g.output && !g.output.startsWith('/') ? path.join(cwd, g.output) : null,
  }))

  generators.sort((a, b) => (a.provider === 'photogen' ? 1 : b.provider === 'photogen' ? -1 : a.name < b.name ? -1 : 1))

  return generators.map((definition, index) => {
    const generator = definitions[definition.provider]
    if (!generator) {
      throw new Error(
        `Unknown generator provider ${definition.provider} for generator ${definition.name} defined in ${path.join(
          cwd,
          'datamodel.prisma',
        )}`,
      )
    }

    const otherGenerators = generators.filter((g, i) => i !== index)

    return {
      prettyName: generator.definition.prettyName,
      generate: () =>
        generateInThread({
          packagePath: generator.packagePath,
          config: {
            cwd,
            generator: definition,
            otherGenerators,
          },
        }),
    } as CompiledGeneratorDefinition
  })
}
