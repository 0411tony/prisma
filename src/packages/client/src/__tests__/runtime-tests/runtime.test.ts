import fs from 'fs'
import path from 'path'
import { generateInFolder } from '../../utils/generateInFolder'
import { promisify } from 'util'
import rimraf from 'rimraf'
import stripAnsi from 'strip-ansi'
const del = promisify(rimraf)

jest.setTimeout(35000)

// TODO: Figure out the problem with debug

process.setMaxListeners(100)

let subDirs = getSubDirs(__dirname)
const folderFilter = process.argv.length > 3 ? process.argv[3] : null
if (folderFilter) {
  subDirs = subDirs.filter((dir) => dir.includes(folderFilter))
  console.log(
    `As ${folderFilter} is provided, only ${subDirs.join(
      ', ',
    )} is being tested`,
  )
}

console.log({ subDirs })

for (const dir of subDirs) {
  const nodeModules = path.join(dir, 'node_modules')
  const testName = path.basename(dir)
  const shouldSucceed = shouldTestSucceed(dir)

  const testTitle = `${testName} example should${
    shouldSucceed ? '' : ' not'
  } succeed`
  test.concurrent(testTitle, async () => {
    if (fs.existsSync(nodeModules)) {
      await del(nodeModules)
    }
    const envVars = getEnvVars(dir)
    process.env = { ...process.env, ...envVars }

    await generateInFolder({
      projectDir: dir,
      useLocalRuntime: false,
      transpile: true,
    })

    if (envVars) {
      for (const key of Object.keys(envVars)) {
        delete process.env[key]
      }
    }

    const filePath = path.join(dir, 'index.js')
    const fn = require(filePath)

    if (shouldSucceed) {
      expect((await fn()) || 'success').toMatchSnapshot(testTitle)
    } else {
      try {
        await fn()
      } catch (e) {
        // https://regex101.com/r/GPVRYg/1/
        // remove the paths, so the tests can succeed on any machine
        expect(
          stripAnsi(e.message).replace(/(\/[\/\S+]+)/gm, ''),
        ).toMatchSnapshot(testTitle)
      }
    }
  })
}

function getSubDirs(dir: string): string[] {
  const files = fs.readdirSync(dir)
  return files
    .map((file) => path.join(dir, file))
    .filter(
      (file) =>
        fs.lstatSync(file).isDirectory() && !file.includes('__snapshots__'),
    )
}

function getEnvVars(dir: string): { [key: string]: string } | undefined {
  const envPath = path.join(dir, 'env.json')
  if (fs.existsSync(envPath)) {
    return JSON.parse(fs.readFileSync(envPath, 'utf-8'))
  }

  return undefined
}

function shouldTestSucceed(dir: string): boolean {
  const manifestPath = path.join(dir, 'manifest.json')
  if (!fs.existsSync(manifestPath)) {
    throw new Error(`Runtime Test dir ${dir} needs a manifest.json`)
  }
  const manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf-8'))

  return manifest.shouldSucceed
}
