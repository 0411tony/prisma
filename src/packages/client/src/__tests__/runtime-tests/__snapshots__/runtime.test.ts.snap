// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blog example should succeed: blog example should succeed 1`] = `"success"`;

exports[`blog-env example should not succeed: blog-env example should not succeed 1`] = `
"error: Environment variable not found: SQLITE_URL.
  -->  schema.prisma:3
   | 
 2 |   provider = \\"sqlite\\"
 3 |   url      = env(\\"SQLITE_URL\\")
   | 

Validation Error Count: 1"
`;

exports[`blog-env-postgresql example should succeed: blog-env-postgresql example should succeed 1`] = `"success"`;

exports[`chmod example should succeed: chmod example should succeed 1`] = `"success"`;

exports[`corruption example should not succeed: corruption example should not succeed 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation in



  Query engine process killed with signal SIGKILL for unknown reason.
Make sure that the engine binary at  is not corrupt.
"
`;

exports[`enums example should succeed: enums example should succeed 1`] = `"success"`;

exports[`hooks example should succeed: hooks example should succeed 1`] = `"success"`;

exports[`incorrect-column-type example should not succeed: incorrect-column-type example should not succeed 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation in



  Attempted to serialize scalar '123' with incompatible type 'String'"
`;

exports[`missing-binary example should not succeed: missing-binary example should not succeed 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation in



  Query engine binary for current platform \\"darwin\\" could not be found.
This probably happens, because you built Prisma Client on a different platform.
(Prisma Client looked in \\"

Files in 

  Dataloader.d.ts
  browser-chalk.d.ts
  browser-terminal-link.d.ts
  browser.d.ts
  dmmf-types.d.ts
  dmmf.d.ts
  error-types.d.ts
  externalToInternalDmmf.d.ts
  getLogLevel.d.ts
  getPrismaClient.d.ts
  highlight
  index.d.ts
  index.js
  index.js.map
  mergeBy.d.ts
  query.d.ts
  transformDmmf.d.ts
  utils
  visit.d.ts


To solve this problem, add the platform \\"darwin\\" to the \\"generator\\" block in the \\"schema.prisma\\" file:
generator client {
  provider      = \\"prisma-client-js\\"
  binaryTargets = [\\"native\\"]
}

Then run \\"prisma generate\\" for your changes to take effect.
Read more about deploying Prisma Client: https:"
`;

exports[`missing-binary-native example should not succeed: missing-binary-native example should not succeed 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation in



  Query engine binary for current platform \\"darwin\\" could not be found.
This probably happens, because you built Prisma Client on a different platform.
(Prisma Client looked in \\"

Files in 

  Dataloader.d.ts
  browser-chalk.d.ts
  browser-terminal-link.d.ts
  browser.d.ts
  dmmf-types.d.ts
  dmmf.d.ts
  error-types.d.ts
  externalToInternalDmmf.d.ts
  getLogLevel.d.ts
  getPrismaClient.d.ts
  highlight
  index.d.ts
  index.js
  index.js.map
  mergeBy.d.ts
  query.d.ts
  transformDmmf.d.ts
  utils
  visit.d.ts

You already added the platform \\"native\\" to the \\"generator\\" block
in the \\"schema.prisma\\" file as described in https:
but something went wrong. That's suboptimal.

Please create an issue at https:"
`;

exports[`missing-column example should not succeed: missing-column example should not succeed 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation in



  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\\"no such column: dev.User.name\\"))) })"
`;

exports[`missing-table example should not succeed: missing-table example should not succeed 1`] = `
"
Invalid \`prisma.user.findMany()\` invocation in



  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\\"no such table: dev.User\\"))) })"
`;

exports[`new-line example should succeed: new-line example should succeed 1`] = `"success"`;

exports[`restart example should succeed: restart example should succeed 1`] = `"success"`;

exports[`transaction example should succeed: transaction example should succeed 1`] = `"success"`;
